auth/shared/src/main/scala/io/udash/auth/Permission.scala

```diff
@@ -0,0 +1,34 @@
+package io.udash.auth
+
+import scala.language.implicitConversions
+
+/**
+  * Base class for permissions used with PermissionControl.
+  * Any subclasses have to be defined in shared project as case classes/objects.
```

Comment:
This can be partially enforced by a self type annotation for Product.

Reply:
I'm not gonna lie: I copied this trait from our internal code and I don't have any idea why subclasses have to case class/object. ;)

Reply:
I do, I wrote it :D That was back when GenCodec was the bad guy and we used the automatic mode, so a lot of bad stuff happened when they weren't.

Reply:
I'm going to remove this comment, because there is no such requirement from `Permission` perspective.